NDTensors/src/lib/UnallocatedArrays/src/unallocatedfill.jl

```diff
@@ -0,0 +1,67 @@
+using FillArrays: FillArrays, AbstractFill, Fill, broadcasted_fill, kron_fill, mult_fill
+using NDTensors.SetParameters: Position, set_parameters
+
+struct UnallocatedFill{ElT,N,Axes,Alloc} <: AbstractFill{ElT,N,Axes}
+  f::Fill{ElT,N,Axes}
+  alloc::Alloc
+end
+
+function UnallocatedFill{ElT,N,Axes}(f::Fill, alloc::Type) where {ElT,N,Axes}
+  return new{ElT,N,Axes,Type{alloc}}(f, alloc)
+end
+
+function UnallocatedFill{ElT,N}(f::Fill, alloc) where {ElT,N}
+  return UnallocatedFill{ElT,N,typeof(axes(f))}(f, alloc)
+end
+
+function UnallocatedFill{ElT}(f::Fill, alloc) where {ElT}
+  return UnallocatedFill{ElT,ndims(f)}(f, alloc)
+end
+
+set_alloctype(f::Fill, alloc::Type) = UnallocatedFill(f, alloc)
+
+Base.parent(F::UnallocatedFill) = F.f
+
+Base.convert(::Type{<:UnallocatedFill}, A::UnallocatedFill) = A
+
+#############################################
+# Arithmatic
+
+# mult_fill(a, b, val, ax) = Fill(val, ax)
+function FillArrays.mult_fill(a::UnallocatedFill, b, val, ax)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+FillArrays.mult_fill(a, b::UnallocatedFill, val, ax) = mult_fill(b, a, val, ax)
+function FillArrays.mult_fill(a::UnallocatedFill, b::UnallocatedFill, val, ax)
+  @assert alloctype(a) == alloctype(b)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+
+function FillArrays.broadcasted_fill(f, a::UnallocatedFill, val, ax)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+function FillArrays.broadcasted_fill(f, a::UnallocatedFill, b::UnallocatedFill, val, ax)
+  @assert alloctype(a) == alloctype(b)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+
+function FillArrays.broadcasted_fill(f, a::UnallocatedFill, b, val, ax)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+function FillArrays.broadcasted_fill(f, a, b::UnallocatedFill, val, ax)
+  return broadcasted_fill(f, b, a, val, ax)
+end
+
+function FillArrays.kron_fill(a::UnallocatedFill, b::UnallocatedFill, val, ax)
+  @assert alloctype(a) == alloctype(b)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+
+Base.:+(A::UnallocatedFill, B::UnallocatedFill) = A .+ B
+
+function Base.Broadcast.broadcasted(
+  ::Base.Broadcast.DefaultArrayStyle, op, r::UnallocatedFill
+)
```

Comment:
Is this definition necessary given that `broadcasted_fill` is defined? I thought `broadcasted_fill` was the only thing we would need for broadcasting.

Reply:
This definition was related to our attempt to do something like `Complex.(F)` or `Float32.(F)` where F is an `UnallocatedFill`. Without it the function `complex` will convert the `UnallocatedFill` to a `Fill`. 
I did find one other issue: This function does not help for converting `UnallocatedZeros` so if I rewrite the function to look like 
```
function Base.Broadcast.broadcasted(
  ::Base.Broadcast.DefaultArrayStyle, op, r::UnallocatedArray
)
  f = op.(parent(r))
  return set_alloctype(f, set_parameters(alloctype(r), Position{1}(), eltype(f)))
end
```
It should work for both `Unallocated` types. However, the function written above convert `UnallocatedZeros` to  `UnallocatedFill` because of the line `f = op.(parent(r))`.  I.e.
```
julia> z = Zeros{Float64}(2,3)
julia> Float32.(z)
2×3 Fill{Float32}, with entries equal to 0.0f0
```
Please let me know what you'r thought are regarding this function. Potentially I could rewrite `complex` (as it is the only function affected by this code) so it does not use broadcasting.

Reply:
I see, thanks for the explanation. I do see that for unary broadcasting operations it is hardcoded to return `Fill`, `Zeros`, etc. instead of calling `broadcast_fill`, `broadcast_zeros`, etc. (https://github.com/JuliaArrays/FillArrays.jl/blob/v1.9.3/src/fillbroadcast.jl#L76-L91) so I see why you would need specialized definitions like `Broadcast.broadcasted(::Broadcast.DefaultArrayStyle, op, f::UnallocatedFill)` and `Broadcast.broadcasted(::Broadcast.DefaultArrayStyle, op, f::UnallocatedZeros)`.

In terms of expressions like `Complex.(f)`, `Float32.(f)`, `complex.(f)`, etc. the tricky part is that we, or `FillArrays`, need to tell Julia's broadcasting system that those operations preserve zero values (as opposed to something like `(x -> x + 1).(f)`, i.e. adding one to every entry).

`FillArrays` takes the strategy of hardcoding which functions preserve zero values:
https://github.com/JuliaArrays/FillArrays.jl/blob/v1.9.3/src/fillbroadcast.jl#L86-L91
https://github.com/JuliaArrays/FillArrays.jl/blob/v1.9.3/src/fillbroadcast.jl#L125-L167
but clearly that isn't comprehensive.

We could make definitions like that ourselves just for `UnallocatedZeros` for operations that `FillArrays` is missing, for example:
```julia
Broadcast.broadcasted( ::Broadcast.DefaultArrayStyle, f::typeof(complex), a::UnallocatedZeros) = FillArrays.broadcasted_zeros(f, a, eltype(a), axes(a))
```

Reply:
As discussed I used the trait based version of `Broadcast` which you have above and defined `IsZeroPreserving()` to be constructed for `::typeof(Complex)` and `::typeof(Real)` I also added unit tests for the case of `(x -> x .+ 1).(UnallocatedZero)` and `UnallocatedFill` to ensure that the types are preserved.